### PR TITLE
Properly delete all network interfaces of a Pod

### LINKED ISSUE #55: VLAN tagging is not cleared on VFs after Pod termination
**Is this a BUG REPORT or FEATURE REQUEST?**:
bug

**What happened**:
`vlan`, `cidr` and `allocation_pool` attributes are set in a `NetworkType: sriov` DanmNet definition.
If a Pod requests such network(s), the VLAN tagging is set properly on VFs selected by SR-IOV Device Plugin. The problem is that VLAN tagging remains enabled on these VFs after Pod termination.
Subsequent Pods cannot run DPDK applications on these VFs (no runtime error, but packages are not leaving the NIC). We need to remove VLAN tagging manually. 

**What you expected to happen**:
VLAN tagging is cleared on VFs after Pod termination. 

**How to reproduce it**:
```
$ ip link show ens1f0
8: ens1f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether ec:0d:9a:6b:a9:14 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
$ kubectl create -f testpod-sriov-6-double.yaml
pod/testpod-sriov-6-double created
$ ip link show ens1f0
8: ens1f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether ec:0d:9a:6b:a9:14 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
$ kubectl delete -f testpod-sriov-6-double.yaml
pod "testpod-sriov-6-double" deleted
$ ip link show ens1f0
8: ens1f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether ec:0d:9a:6b:a9:14 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
```

**Anything else we need to know?**:
I believe that this behavior is not introduced by #50 , so it does not seem to be a new bug.
Currently it is not clear whether it is a Danm or netlink responsibility. What do you think?

### PR DESCRIPTION
**Pull request by Levovar**
**What type of PR is this?**
bug

**Which issue(s) this PR fixes** *
Fixes #55 (hopefully :) )

### CODE REVIEW

**Comment by TothFerenc** [comment]
It works functional point of view:
```
$ ip link show ens1f0
8: ens1f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether ec:0d:9a:6b:a9:14 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
$ kubectl create -f testpod-sriov-6-double.yaml
pod/testpod-sriov-6-double created
$ ip link show ens1f0
8: ens1f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether ec:0d:9a:6b:a9:14 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, vlan 100, spoof checking off, link-state auto, trust off, query_rss off
$ kubectl delete -f testpod-sriov-6-double.yaml
pod "testpod-sriov-6-double" deleted
$ ip link show ens1f0
8: ens1f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether ec:0d:9a:6b:a9:14 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust off, query_rss off
$
```
But these lines are logged:
```
2019/03/08 13:17:18 CNI ADD invoked with: ns:default PID:testpod-sriov-6-double CID: dfef2492e0f583ea23edb1fc0d2dd17b8a15d015d21d26f6e401c4af15d9b67e
2019/03/08 13:18:02 CNI DEL invoked with: ns:default PID:testpod-sriov-6-double CID: dfef2492e0f583ea23edb1fc0d2dd17b8a15d015d21d26f6e401c4af15d9b67e
2019/03/08 13:18:03 INFO: DEL: Following errors happened during interface deletion:CNI operation for network:sriov-danm2 failed with:failed to delete container NIC:Error delegating DEL to CNI plugin:sriov because:unexpected end of JSON input;
CNI operation for network:sriov-danm1 failed with:failed to delete container NIC:Error delegating DEL to CNI plugin:sriov because:unexpected end of JSON input;
CNI operation for network:sriov-danm2 failed with:failed to delete container NIC:Error delegating DEL to CNI plugin:sriov because:unexpected end of JSON input;
CNI operation for network:flannel failed with:failed to delete container NIC:Error delegating DEL to CNI plugin:flannel because:unexpected end of JSON input;
CNI operation for network:sriov-danm1 failed with:failed to delete container NIC:Error delegating DEL to CNI plugin:sriov because:unexpected end of JSON input;
CNI operation for network:sriov-danm1 failed with:failed to delete container NIC:Error delegating DEL to CNI plugin:sriov because:unexpected end of JSON input;
CNI operation for network:sriov-danm2 failed with:failed to delete container NIC:Error delegating DEL to CNI plugin:sriov because:unexpected end of JSON input;
```
Is this something we can ignore or do we need to fix it now?

**Comment by Levovar** [comment]
nah, we are not in the process of providing incomplete fixes :)
I guessed something like this could happen: simply speaking there is no CNI result to return, so some function returns error in some case it shouldnt

let me try again, and let's see. but yes, this is purely cosmetic, does not affect functionality. we never report "error" to kubelet for the whole CNI_DEL operation, even if some subtasks have failed

**Comment by Levovar** [comment]
@TothFerenc try this one

**Comment by TothFerenc** [comment]
Good news! No CNI DEL error messages in the log:
```
2019/03/08 22:27:27 CNI ADD invoked with: ns:default PID:testpod-sriov-6-double CID: 00cb69ec4cc9e9953037ffd59d5e88a63d67db9785af405807a01859e759e225
2019/03/08 22:28:10 CNI DEL invoked with: ns:default PID:testpod-sriov-6-double CID: 00cb69ec4cc9e9953037ffd59d5e88a63d67db9785af405807a01859e759e225
```

**Comment by Levovar** [comment]
finally :D
thanks for the testing!

**Comment by BorosT** [comment]
Hello! could you please issue a new release after this pull request was successfully merged!

**Comment by Levovar** [comment]
> @Levovar, could you please issue a new release after this pull request was successfully merged!

of course :) just did it!
sry for the delay, but I also need to live sometime :)

**Comment by BorosT** [comment]
> 
> 
> > @Levovar, could you please issue a new release after this pull request was successfully merged!
> 
> of course :) just did it!
> sry for the delay, but I also need to live sometime :)

Thank you!
